sdk/core/src/tokio/fs.rs

```diff
@@ -0,0 +1,165 @@
+
+    pub async fn build(mut self) -> crate::Result<FileStream> {
+        let stream_size = self.handle.metadata().await?.len();
+
+        let buffer_size = self.buffer_size.unwrap_or(DEFAULT_BUFFER_SIZE);
+
+        let offset = if let Some(offset) = self.offset {
+            self.handle.seek(SeekFrom::Start(offset)).await?;
+            offset
+        } else {
+            0
+        };
+
+        let block_size = if let Some(block_size) = self.block_size {
+            block_size
+        } else {
+            stream_size - offset
+        };
+
+        let handle = Arc::new(Mutex::new(self.handle.take(block_size)));
+
+        Ok(FileStream {
+            handle,
+            buffer_size,
+            block_size,
+            stream_size,
+            offset,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+#[pin_project::pin_project]
+pub struct FileStream {
+    #[pin]
+    handle: Arc<Mutex<Take<File>>>,
+    pub stream_size: u64,
+    pub block_size: u64,
+    buffer_size: usize,
+    pub offset: u64,
+}
+
+impl FileStream {
+    async fn read(&mut self, slice: &mut [u8]) -> std::io::Result<usize> {
+        let mut handle = self.handle.clone().lock_owned().await;
+        handle.read(slice).await
+    }
+
+    /// Resets the number of bytes that will be read from this instance to the
+    /// `stream_size`
+    ///
+    /// This is useful if you want to read the stream in multiple blocks
+    pub async fn next_block(&mut self) -> crate::Result<()> {
+        log::info!("setting limit to {}", self.block_size);
+        let mut handle = self.handle.clone().lock_owned().await;
+        {
+            let inner = handle.get_mut();
+            self.offset = inner.stream_position().await?;
+        }
+        handle.set_limit(self.block_size);
+        Ok(())
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl SeekableStream for FileStream {
+    /// Seek to the specified offset into the file and reset the number of bytes to read
+    ///
+    /// This is useful upon encountering an error to reset the stream to the last
+    async fn reset(&mut self) -> crate::Result<()> {
+        log::info!(
+            "resetting stream to offset {} and limit to {}",
+            self.offset,
+            self.block_size
+        );
+        let mut handle = self.handle.clone().lock_owned().await;
+        {
+            let inner = handle.get_mut();
+            inner.seek(SeekFrom::Start(self.offset)).await?;
+        }
+        handle.set_limit(self.block_size);
+        Ok(())
+    }
+
+    fn len(&self) -> usize {
+        log::info!(
+            "stream len:  {} - {} ... {}",
+            self.stream_size,
+            self.offset,
+            self.block_size
+        );
+        min(self.stream_size - self.offset, self.block_size) as usize
+    }
+
+    /*
+    fn buffer_size(&self) -> usize {
+        self.buffer_size
+    }
+     */
+}
```

Comment:
Did you intend to leave this comment in?

Reply:
This is an artifact from trying to split up the work into multiple PRs.  In this case, #1359.